packages/message-relayer/src/relay-tx.ts

```diff
@@ -0,0 +1,397 @@
+    return null
+  } else if (!isEventHi(batchEvent, l2TransactionIndex)) {
+    // Upper bound is not too low and also not too high. This means the upper bound event is the
+    // one we're looking for! Return it.
+    return batchEvent
+  }
+
+  // Binary search to find the right event. The above checks will guarantee that the event does
+  // exist and that we'll find it during this search.
+  while (lowerBound < upperBound) {
+    const middleOfBounds = Math.floor((lowerBound + upperBound) / 2)
+    batchEvent = await getStateBatchAppendedEventByBatchIndex(middleOfBounds)
+
+    if (isEventHi(batchEvent, l2TransactionIndex)) {
+      upperBound = middleOfBounds
+    } else if (isEventLo(batchEvent, l2TransactionIndex)) {
+      lowerBound = middleOfBounds
+    } else {
+      break
+    }
+  }
+
+  return batchEvent
+}
+
+/**
+ * Finds the full state root batch associated with a given transaction index.
+ * @param l1RpcProvider L1 RPC provider.
+ * @param l1StateCommitmentChainAddress Address of the L1StateCommitmentChain.
+ * @param l2TransactionIndex Index of the L2 transaction to find a state root batch for.
+ * @returns State root batch associated with the given transaction index or null if no state root
+ * batch exists.
+ */
+export const getStateRootBatchByTransactionIndex = async (
+  l1RpcProvider: ethers.providers.JsonRpcProvider,
+  l1StateCommitmentChainAddress: string,
+  l2TransactionIndex: number
+): Promise<StateRootBatch | null> => {
+  const l1StateCommitmentChain = new ethers.Contract(
+    l1StateCommitmentChainAddress,
+    getContractInterface('OVM_StateCommitmentChain'),
+    l1RpcProvider
+  )
+
+  const stateBatchAppendedEvent = await getStateBatchAppendedEventByTransactionIndex(
+    l1RpcProvider,
+    l1StateCommitmentChainAddress,
+    l2TransactionIndex
+  )
+  if (stateBatchAppendedEvent === null) {
+    return null
+  }
+
+  const stateBatchTransaction = await stateBatchAppendedEvent.getTransaction()
+  const [stateRoots] = l1StateCommitmentChain.interface.decodeFunctionData(
+    'appendStateBatch',
+    stateBatchTransaction.data
+  )
+
+  return {
+    header: {
+      batchIndex: stateBatchAppendedEvent.args._batchIndex,
+      batchRoot: stateBatchAppendedEvent.args._batchRoot,
+      batchSize: stateBatchAppendedEvent.args._batchSize,
+      prevTotalElements: stateBatchAppendedEvent.args._prevTotalElements,
+      extraData: stateBatchAppendedEvent.args._extraData,
+    },
+    stateRoots,
+  }
+}
+
+/**
+ * Generates a Merkle proof (using the particular scheme we use within Lib_MerkleTree).
+ * @param leaves Leaves of the merkle tree.
+ * @param index Index to generate a proof for.
+ * @returns Merkle proof sibling leaves, as hex strings.
+ */
+const getMerkleTreeProof = (leaves: string[], index: number): string[] => {
+  const parsedLeaves = []
+  for (let i = 0; i < Math.pow(2, Math.ceil(Math.log2(leaves.length))); i++) {
```

Comment:
Consider moving `Math.ceil(...)` to another line so its only evaluated once

Reply:
Moved and added some comments to clarify what's happening here.